File: 2020/day01/src/main.rs
use std::collections::HashMap;
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;

fn main() {
    const TARGET_NUMBER: i32 = 2020;
    let mut numbers = Vec::new();
    let mut adds_to = HashMap::new();
    let mut cmd_line = std::env::args();
    let file_to_read = if cmd_line.len() > 1 {
        cmd_line.nth(1).unwrap()
    } else {
        "sample_input.txt".to_string()
    };
    if let Ok(lines) = read_lines(file_to_read) {
        for line in lines {
            if let Ok(ip) = line {
                if let Ok(a) = ip.parse::<i32>() {
                    match adds_to.get(&a) {
                        Some(other) => {
                            let b = numbers[*other];
                            println!("Found {} and {}, mult to {}", a, b, a * b);
                        }
                        None => {}
                    }

                    let index: usize = numbers.len();
                    numbers.push(a);
                    let inv = TARGET_NUMBER - a;
                    adds_to.insert(inv, index);
                }
            }
        }
    }
    let n = numbers.len();
    for i in 0..n {
        for j in i + 1..n {
            let a = numbers[i];
            let b = numbers[j];
            let temp = a + b;
            match adds_to.get(&temp) {
                Some(k) => {
                    if *k < j {
                        continue;
                    }
                    let c = numbers[*k];
                    println!("Found {}, {}, and {}, mult to {}", a, b, c, a * b * c);
                }
                None => {
                    continue;
                }
            }
        }
    }
}

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

File: 2020/day01/main.py
import sys

if __name__ == "__main__":
    target_num = 2020
    adds_to = {}
    numbers: list[int] = []
    try:
        fname = sys.argv[1]
    except:
        fname = "sample_input.txt"
    with open(fname, "r") as f:
        for line in f:
            try:
                n = int(line)
            except:
                continue
            inv = target_num - n
            if n in adds_to:
                print(n, inv, n * inv)
            index = len(numbers)
            adds_to[inv] = index
            numbers.append(n)
    for i in range(len(numbers)):
        for j in range(i + 1, len(numbers)):
            a = numbers[i]
            b = numbers[j]
            k = adds_to.get(a + b)
            if k:
                if k < j:
                    continue
                c = numbers[k]
                print(a, b, c, a * b * c)
